src/shims/io.rs

```diff
@@ -0,0 +1,213 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+
+use rustc::ty::layout::Size;
+
+use crate::stacked_borrows::Tag;
+use crate::*;
+
+pub struct FileHandle {
+    file: File,
+    flag: i32,
+}
+
+pub struct FileHandler {
+    handles: HashMap<i32, FileHandle>,
+    low: i32,
+}
+
+impl Default for FileHandler {
+    fn default() -> Self {
+        FileHandler {
+            handles: Default::default(),
+            // 0, 1 and 2 are reserved for stdin, stdout and stderr
+            low: 3,
+        }
+    }
+}
+
+impl<'mir, 'tcx> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    fn open(
+        &mut self,
+        path_op: OpTy<'tcx, Tag>,
+        flag_op: OpTy<'tcx, Tag>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let flag = this.read_scalar(flag_op)?.to_i32()?;
+
+        if flag != this.eval_libc_i32("O_RDONLY")? && flag != this.eval_libc_i32("O_CLOEXEC")? {
+            throw_unsup_format!("Unsupported flag {:#x}", flag);
+        }
+
+        let path_bytes = this
+            .memory()
+            .read_c_str(this.read_scalar(path_op)?.not_undef()?)?;
+        let path = std::str::from_utf8(path_bytes)
+            .map_err(|_| err_unsup_format!("{:?} is not a valid utf-8 string", path_bytes))?;
+        let fd = File::open(path).map(|file| {
+            let mut fh = &mut this.machine.file_handler;
+            fh.low += 1;
+            fh.handles.insert(fh.low, FileHandle { file, flag });
+            fh.low
+        });
+
+        this.consume_result(fd)
+    }
+
+    fn fcntl(
+        &mut self,
+        fd_op: OpTy<'tcx, Tag>,
+        cmd_op: OpTy<'tcx, Tag>,
+        arg_op: Option<OpTy<'tcx, Tag>>,
+    ) -> InterpResult<'tcx, i32> {
+        let this = self.eval_context_mut();
+
+        if !this.machine.communicate {
+            throw_unsup_format!("`open` not available when isolation is enabled")
+        }
+
+        let fd = this.read_scalar(fd_op)?.to_i32()?;
+        let cmd = this.read_scalar(cmd_op)?.to_i32()?;
+
+        if cmd == this.eval_libc_i32("F_SETFD")? {
+            // This does not affect the file itself. Certain flags might require changing the file
+            // or the way it is accessed somehow.
+            let flag = this.read_scalar(arg_op.unwrap())?.to_i32()?;
+            // The only usage of this in stdlib at the moment is to enable the `FD_CLOEXEC` flag.
+            let fd_cloexec = this.eval_libc_i32("FD_CLOEXEC")?;
+            if let Some(FileHandle { flag: old_flag, .. }) =
+                this.machine.file_handler.handles.get_mut(&fd)
+            {
+                if flag ^ *old_flag == fd_cloexec {
```

Comment:
Why `^`? That's a rare sight in flag handling code.

What I expected would have been something like `if flag & fd_cloexec != 0`, followed by `*old_flag |= fd_cloexec`.

Reply:
I thought this was a simple enough way of checking that the only change done to `flag`/`old_flag` was introducing `fd_cloexec`. For example, if you're in write mode you'll be in the following situation
`old_flag = 1`
`flag = fd_cloexec & 1`
So `flag == fd_cloexec` would be `false`.

I could something like  `flag & fd_cloexec != 0 && old_flag & fd_cloexec == 0` if you want.

Reply:
Oh I see. Yeah that should work but when you write obfuscated code like this, please explain in a comment what is happening. :)

Reply:
Yeah sorry I was in the same mood as with the escaping strings yesterday so I was cursing and reading man until it worked. Let me add the comment

Reply:
https://github.com/rust-lang/miri/pull/983/commits/1c64f29811db779bb202c3450c60c227fb0234bb